plutus-emulator/src/Wallet/Emulator/Types.hs

```diff
@@ -504,13 +506,22 @@ liftMockWallet wallet act = do
     let walletState = fromMaybe (emptyWalletState wallet) $ Map.lookup wallet $ _walletStates emState
         ((out, newState), (msgs, txns)) = runWriter $ runStateT (runExceptT (runMockWallet act)) walletState
         events = (TxnSubmit . hashTx <$> txns) ++ (WalletInfo wallet <$> getWalletLog msgs)
-    put emState {
-        _txPool = txns ++ _txPool emState,
-        _walletStates = Map.insert wallet newState $ _walletStates emState,
-        _emulatorLog = events ++ _emulatorLog emState
-        }
+    liftNodeClient $ mapM_ NC.publishTx txns
+    modify $ set walletStates (Map.insert wallet newState $ _walletStates emState)
+           . set emulatorLog  (events ++ _emulatorLog emState)
     pure (txns, out)
 
+liftNodeClient :: MonadState EmulatorState m => State NC.NodeClientState a -> m a
```

Comment:
I wanted to say that surely we just have a `MonadState ChainState m` constraint or something, but of course MTL doesn't let us have multiple `MonadState` constraints, and "zooming" in on a bit of state is a pain unless we do some kind of `MonadState s m, Has s ChainState` dance...

Reply:
I feel this pain constantly. I will go dancing :)

Reply:
Looking a this again, I think the dance is indeed probably the way to go, since we do have multiple bits of cooperating state here.